### PR TITLE
feat(lab): scaffolds a basic UI test

### PR DESCRIPTION
**Pull request by arein**
# Breaking Changes

N/A

# Changes

- feat(lab): scaffolds a basic UI test

Next: https://github.com/WalletConnect/web3modal/issues/1582

# Associated Issues

closes #...

### CODE REVIEW

File: apps/laboratory/package.json [hunk 1]
```diff
@@ -5,7 +5,11 @@
   "scripts": {
     "dev:laboratory": "next dev",
     "build:laboratory": "next build",
-    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
+    "lint": "eslint . --ext .js,.jsx,.ts,.tsx",
+    "playwright:start": "next dev",
```
Comment (0xAsimetriq) [inline apps/laboratory/package.json:9]: If tests are for lab, this should be `npm run laboratory` to start in in dev mode. Optionally, you could combine with a fresh build as well, so `turbo run build; npm run laboratory`

File: apps/laboratory/tests/shared/constants/index.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export const LOCAL_LABS_URL = 'https://lab.web3modal.com/library/wagmi/'
```
Comment (0xAsimetriq) [inline apps/laboratory/tests/shared/constants/index.ts:1]: Should this be localhot:3000 or should variable be re-named?

**Review by 0xAsimetriq — COMMENTED** [review]
Left few minor suggestions. I think at the moment this is testing against lab.web3modal.com and not branch code?

**Review by chris13524 — CHANGES_REQUESTED** [review]
Use [`webServer.command`](https://playwright.dev/docs/test-configuration) to spin up a dev version of the tests when running. Similar to how it was done for [V2](https://github.com/WalletConnect/web3modal/blob/2a0e946f9c848c25d6c9cf3c53ab4a04ed372345/laboratory/playwright.config.ts#L69)

**Comment by arein** [comment]
> Use [`webServer.command`](https://playwright.dev/docs/test-configuration) to spin up a dev version of the tests when running. Similar to how it was done for [V2](https://github.com/WalletConnect/web3modal/blob/2a0e946f9c848c25d6c9cf3c53ab4a04ed372345/laboratory/playwright.config.ts#L69)

thank you was wondering how this works and missed this when I copied from V2

File: .github/workflows/ui_tests.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      - name: build
+        run: npm run build
+      - name: Install Playwright Browsers
+        working-directory: ./apps/laboratory/
+        run: npm run playwright:install
+      - name: Run Playwright tests
+        env:
+          NEXT_PUBLIC_PROJECT_ID: ${{ secrets.NEXT_PUBLIC_PROJECT_ID }}
+        working-directory: ./apps/laboratory/
+        run: npm run playwright:test:${{ matrix.project }}
```
Comment (chris13524) [inline .github/workflows/ui_tests.yml:51]: We should get rid of this matrix and use Playwright's built-in parallelization which will test against all browsers at once. This will save build minutes building the same thing twice.

```suggestion
        run: npm run playwright:test
```

**Review by chris13524 — CHANGES_REQUESTED** [review]

File: apps/laboratory/package.json (hunk 1 above)
Comment (chris13524) [reply apps/laboratory/package.json]: What's the difference between `npm run build` and `turbo run build`?

File: apps/laboratory/package.json [hunk 4]
```diff
@@ -5,7 +5,11 @@
   "scripts": {
     "dev:laboratory": "next dev",
     "build:laboratory": "next build",
-    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
+    "lint": "eslint . --ext .js,.jsx,.ts,.tsx",
+    "playwright:start": "turbo run build; npm run dev:laboratory",
```
Comment (chris13524) [inline apps/laboratory/package.json:9]: Since I don't have `turbo` installed, I can't run `npm run playwright:test` locally. We should make a `npm run turbo-build -> turbo run build` script and call that here.

**Review by chris13524 — CHANGES_REQUESTED** [review]
1 more thing 🙏

File: apps/laboratory/package.json (hunk 4 above)
Comment (arein) [reply apps/laboratory/package.json:9]: I think I can remove the `turbo run build` from the command. Would that address your concern?

Comment (0xAsimetriq) [reply apps/laboratory/package.json:9]: You can run `npm run build` instead

Comment (chris13524) [reply apps/laboratory/package.json]: @arein the script is broken now; you can't run `npm run build` in the lab directory; that script does not exist. `turbo run build` worked because it recurses into the parent directory.

Comment (0xAsimetriq) [reply apps/laboratory/package.json]: You are right actually. For everything to work properly / build packages etc. Set up and commands should be ran from the monorepo root. Similarly something like `npm run laboratory` should be ran from there. Doing otherwise i.e. running from individual packages could have unexpected side effects probably

Comment (arein) [reply apps/laboratory/package.json]: ok fixed

**Review by chris13524 — APPROVED** [review]

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]
This sets everything up nicely, but does not include new approach we discussed with Jon, Sven and others for v3 testing.

Due to fast moving pace and changes, we'd need to spend substantial amount of time on re-updating tests if they are based on granular checks i.e. like you check for `All Wallets`.

Road forward with v3 that will keep us moving fast while providing good feedback on breaking something is through playwrights visual diff tests

More context here https://playwright.dev/docs/test-snapshots

TL;DR these are extremely cheap to set up dev time wise for existing and upcoming features while providing really good feedback on breaking changes i.e. if something is off, pixels will be off as well (missing notification. modal not open etc.). Updating these will be easy as well, i.e. if something changed and it was expected, we just update reference screenshot.

@arein Could you please switch from using "All Wallets" detection in the test, to one using screenshot diffing i.e. snap of open modal.

File: .github/workflows/ui_tests.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Install Playwright Browsers
+        working-directory: ./apps/laboratory/
+        run: npm run playwright:install
+      - name: Run Playwright tests
+        env:
+          NEXT_PUBLIC_PROJECT_ID: ${{ secrets.NEXT_PUBLIC_PROJECT_ID }}
+        working-directory: ./apps/laboratory/
+        run: npm run playwright:test
+      - uses: actions/upload-artifact@v3
+        if: always()
```
[the remark this answers is not in the document]
Comment (chris13524) [reply .github/workflows/ui_tests.yml:49]: `if: always()` I believe allows it to still upload failure results e.g. screenshots when we set those up

[the remark this answers is not in the document]
Comment (arein) [reply .github/workflows/ui_tests.yml:49]: yes

**Comment by arein** [comment]
> You are right actually. For everything to work properly / build packages etc. Set up and commands should be ran from the monorepo root. Similarly something like npm run laboratory should be ran from there. Doing otherwise i.e. running from individual packages could have unexpected side effects probably

What's actually requested?
As a developer I must be able to run e2e tests from monorepo root?

> This sets everything up nicely, but does not include new approach we discussed with Jon, Sven and others for v3 testing.
Due to fast moving pace and changes, we'd need to spend substantial amount of time on re-updating tests if they are based on granular checks i.e. like you check for All Wallets.
Road forward with v3 that will keep us moving fast while providing good feedback on breaking something is through playwrights visual diff tests
More context here https://playwright.dev/docs/test-snapshots
TL;DR these are extremely cheap to set up dev time wise for existing and upcoming features while providing really good feedback on breaking changes i.e. if something is off, pixels will be off as well (missing notification. modal not open etc.). Updating these will be easy as well, i.e. if something changed and it was expected, we just update reference screenshot.
@arein Could you please switch from using "All Wallets" detection in the test, to one using screenshot diffing i.e. snap of open modal.

I'm scaffolding everything here @0xAsimetriq specifically to avoid comments like these. I am not actually testing anything here. The `All Wallets` is just a placeholder tests. I don't see how this setup cannot be used later for actual diffs

File: apps/laboratory/package.json [hunk 6]
```diff
@@ -5,7 +5,12 @@
   "scripts": {
     "dev:laboratory": "next dev",
     "build:laboratory": "next build",
-    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
+    "build:all": "turbo run build",
```
Comment (0xAsimetriq) [inline apps/laboratory/package.json:8]: Turborepo and monorepo wide commands should not be ran from here. There is global `npm run build` command that does exactly this in the root, can that be reused? It also sets up clean builds properly for all packages.

You will likely need to move playwright commands and test dependencies to monorepo root as well and just target tests inside laboratory folder.

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]

**Comment by chris13524** [comment]
> Due to fast moving pace and changes, we'd need to spend substantial amount of time on re-updating tests if they are based on granular checks i.e. like you check for All Wallets.
> 
> Road forward with v3 that will keep us moving fast while providing good feedback on breaking something is through playwrights visual diff tests

Taking full advantage of Playwright will involve a lot more than screenshots; every edge case should have a test case and needing to update something trivial like "All Wallets" will be the least of your concerns when changing stuff around.

**Comment by 0xAsimetriq** [comment]
> > You are right actually. For everything to work properly / build packages etc. Set up and commands should be ran from the monorepo root. Similarly something like npm run laboratory should be ran from there. Doing otherwise i.e. running from individual packages could have unexpected side effects probably
> 
> What's actually requested? As a developer I must be able to run e2e tests from monorepo root?
> 
> > This sets everything up nicely, but does not include new approach we discussed with Jon, Sven and others for v3 testing.
> > Due to fast moving pace and changes, we'd need to spend substantial amount of time on re-updating tests if they are based on granular checks i.e. like you check for All Wallets.
> > Road forward with v3 that will keep us moving fast while providing good feedback on breaking something is through playwrights visual diff tests
> > More context here https://playwright.dev/docs/test-snapshots
> > TL;DR these are extremely cheap to set up dev time wise for existing and upcoming features while providing really good feedback on breaking changes i.e. if something is off, pixels will be off as well (missing notification. modal not open etc.). Updating these will be easy as well, i.e. if something changed and it was expected, we just update reference screenshot.
> > @arein Could you please switch from using "All Wallets" detection in the test, to one using screenshot diffing i.e. snap of open modal.
> 
> I'm scaffolding everything here @0xAsimetriq specifically to avoid comments like these. I am not actually testing anything here. The `All Wallets` is just a placeholder tests. I don't see how this setup cannot be used later for actual diffs

So for Q1, right now there seems to be a `build:all` setup in laboratory using global turborepo command. This should not be the case case, build depends on all packages, so it should be only ran and set up in root of the monorepo. I left separate suggestion about moving playwright tests to root as well.

**Comment by 0xAsimetriq** [comment]
> > Due to fast moving pace and changes, we'd need to spend substantial amount of time on re-updating tests if they are based on granular checks i.e. like you check for All Wallets.
> > Road forward with v3 that will keep us moving fast while providing good feedback on breaking something is through playwrights visual diff tests
> 
> Taking full advantage of Playwright will involve a lot more than screenshots; every edge case should have a test case and needing to update something trivial like "All Wallets" will be the least of your concerns when changing stuff around.

Yes, but remember this is ui product so majority of issues will bubble up to ui, which is why screenshot diffing is great here. I.e if you set an option to disable metamask, all you have to do is screenshot that and it should not be there, if it is, then we made a regression. This applies to almost every config / action in web3modal, if something is wrong it will result in a different visual to what is expected (even text in error notifications for example)

**Comment by chris13524** [comment]
> Yes, but remember this is ui product so majority of issues will bubble up to ui, which is why screenshot diffing is great here. I.e if you set an option to disable metamask, all you have to do is screenshot that and it should not be there, if it is, then we made a regression. This applies to almost every config / action in web3modal, if something is wrong it will result in a different visual to what is expected (even text in error notifications for example)

Yes, but I feel like this is the easy way out and is easier to mess up. There's 1 command that will reset all of the screenshots. So say you have a new feature that adds a new wallet to the list of wallets, and you reset all of the screenshots. You get your new wallet, but you also removed MetaMask without realizing it. Simply the screenshot checks will not detect this problem.

Now since the screenshots are committed and reviewed, hopefully someone will notice that the one called "MetaMask exists" does indeed still contain a MetaMask.

I guess my point is that it's easier to mess up an image than an explicit code removal/modifiction of the MetaMask test case.

I haven't used these snapshots before, so maybe I'll see the light :D
